apps/astarte_data_updater_plant/lib/astarte_data_updater_plant/rpc/impl.ex

```diff
@@ -0,0 +1,172 @@
+        cond do
+          Map.has_key?(data_trigger, :group_name) and not is_nil(data_trigger.group_name) ->
+            {:data_trigger_group, data_trigger.group_name}
+
+          Map.has_key?(data_trigger, :device_id) and not is_nil(data_trigger.device_id) ->
+            {:data_trigger_device, data_trigger.device_id}
+
+          true ->
+            {:all, nil}
+        end
+
+      {:device_trigger,
+       %Astarte.Core.Triggers.SimpleTriggersProtobuf.DeviceTrigger{device_id: device_id}} ->
+        {:device, device_id}
+
+      _ ->
+        {:unknown, nil}
+    end
+  end
+
+  def get_devices_to_notify(%State{} = state, scope) do
+    case scope do
+      {:all, nil} ->
+        state
+        |> fetch_all_devices()
+        |> Enum.map(fn %Device{device_id: device_id, realm: realm} -> {device_id, realm} end)
+
+      {:data_trigger_group, group_name} ->
+        state
+        |> fetch_devices_by_group(group_name)
+        |> Enum.map(fn %Device{device_id: device_id, realm: realm} -> {device_id, realm} end)
+
+      {:device, device_id} ->
+        case fetch_device_by_id(state, device_id) do
+          {:ok, {device_id, realm}} ->
+            [{device_id, realm}]
+
+          {:error, :device_not_found} ->
+            Logger.error("Device with ID #{device_id} not found.")
+            []
+        end
+
+      {:data_trigger_device, device_id} ->
+        case fetch_device_by_id(state, device_id) do
+          {:ok, {device_id, realm}} ->
+            [{device_id, realm}]
+
+          {:error, :device_not_found} ->
+            Logger.error("Device with ID #{device_id} not found.")
+            []
+        end
+
+      _ ->
+        Logger.error("Unknown scope for trigger installation detected: #{inspect(scope)}")
+        []
+    end
+  end
+
+  def retrieve_realms() do
+    realms = Queries.fetch_realms!()
+  end
+
+  def convert_extended_device_id_to_string(extended_device_id) do
+    binary_id = Ecto.UUID.dump!(extended_device_id)
+    string_id = Astarte.Core.Device.encode_device_id(binary_id)
+  end
+
+  def retrieve_connected_devices(realm) do
+    Queries.fetch_connected_devices(realm)
+    |> case do
+      {:ok, devices} ->
+        converted_id_devices =
+          Enum.map(devices, fn %Device{} = device ->
+            %Device{device | device_id: convert_extended_device_id_to_string(device.device_id)}
+          end)
+
+        {:ok, converted_id_devices}
+
+      {:error, reason} ->
+        Logger.error("Failed to fetch connected devices for realm #{realm}: #{inspect(reason)}")
+        {:error, reason}
+    end
+  end
+
+  def fetch_devices_by_group(%State{devices: devices}, group) do
+    devices
+    |> Map.values()
+    |> Enum.filter(fn %Device{groups: groups} -> groups && group in groups end)
+  end
+
+  def fetch_device_by_id(%State{devices: devices}, device_id) do
+    case Map.get(devices, device_id) do
+      nil ->
+        {:error, :device_not_found}
```

Comment:
This is cache invalidation I think, is the device really not connected or some time has passed since the last `Queries.fetch_connected_devices/1` and the device is now connected but not in the cache?

I think the code is as good as is (firts make it work, second make it beautiful), but maybe a `TODO` here to signal that we should check for the cache might be good